cloud/libvirt/actuators/machine/utils/network.go

```diff
@@ -0,0 +1,319 @@
+package libvirt
+
+import (
+	"encoding/xml"
+	"fmt"
+	libvirt "github.com/libvirt/libvirt-go"
+	"github.com/libvirt/libvirt-go-xml"
+	"log"
+	"math/rand"
+	"net"
+)
+
+const (
+	netModeIsolated = "none"
+	netModeNat      = "nat"
+	netModeRoute    = "route"
+	netModeBridge   = "bridge"
+)
+
+// Network interface used to expose a libvirt.Network
+type Network interface {
+	GetXMLDesc(flags libvirt.NetworkXMLFlags) (string, error)
+}
+
+func newDefNetworkfromLibvirt(network Network) (libvirtxml.Network, error) {
+	networkXMLDesc, err := network.GetXMLDesc(0)
+	if err != nil {
+		return libvirtxml.Network{}, fmt.Errorf("Error retrieving libvirt domain XML description: %s", err)
+	}
+	networkDef := libvirtxml.Network{}
+	err = xml.Unmarshal([]byte(networkXMLDesc), &networkDef)
+	if err != nil {
+		return libvirtxml.Network{}, fmt.Errorf("Error reading libvirt network XML description: %s", err)
+	}
+	return networkDef, nil
+}
+
+// HasDHCP checks if the network has a DHCP server managed by libvirt
+func HasDHCP(net libvirtxml.Network) bool {
+	if net.Forward != nil {
```

Comment:
This seems odd - why does the presence of `forward` imply it uses DHCP. Looking at my default network I see:

```console
$ virsh net-dumpxml default
<network>
  <name>default</name>
  <uuid>f374ea25-7007-4f22-a5eb-dbd59ee278f0</uuid>
  <forward mode='nat'>
    <nat>
      <port start='1024' end='65535'/>
    </nat>
  </forward>
  <bridge name='virbr0' stp='on' delay='0'/>
  <mac address='52:54:00:ad:07:2d'/>
  <ip address='192.168.122.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.122.2' end='192.168.122.254'/>
    </dhcp>
  </ip>
</network>
```